locator/abstract_replication_strategy.cc

```diff
@@ -235,14 +235,20 @@ std::unordered_multimap<inet_address, dht::token_range>
 abstract_replication_strategy::get_address_ranges(const token_metadata& tm, can_yield can_yield) const {
     std::unordered_multimap<inet_address, dht::token_range> ret;
     for (auto& t : tm.sorted_tokens()) {
-        dht::token_range_vector r = tm.get_primary_ranges_for(t);
-        auto eps = calculate_natural_endpoints(t, tm, can_yield);
+        dht::token_range_vector r = tm.get_primary_ranges_for(t); // O(1)
+        auto eps = calculate_natural_endpoints(t, tm, can_yield); // O(number of tokens in the ring)
         logger.debug("token={}, primary_range={}, address={}", t, r, eps);
-        for (auto ep : eps) {
+        for (auto ep : eps) { // O(size of the cluster)
             for (auto&& rng : r) {
                 ret.emplace(ep, rng);
             }
         }
+
+        // Each iteration of this outer for loop has O(number of tokens in the ring) complexity so it's fast,
+        // but in total the loop has O((number of tokens in the ring)^2) complexity; yield after each iteration to avoid stalls.
+        if (can_yield) {
+            seastar::thread::maybe_yield();
```

Comment:
We already pass `can_yield` to `calculate_natural_endpoints`.
`network_topology_strategy::calculate_natural_endpoints`, for example, considers yielding based on that (See ba31350239), and so are the other implementations.
Why not add this just to `everywhere_replication_strategy::calculate_natural_endpoints`?

Reply:
@asias @bhalevy 
`calculate_natural_endpoints` is a fast O(n) algorithm, my reasoning was that there is no point in yielding in the middle of computations that are almost instantaneous.
I could add a yield at the beginning of the function like this:
```
inet_address_vector_replica_set everywhere_replication_strategy::calculate_natural_endpoints(const token& search_token, const token_metadata& tm, can_yield) const {
    if (can_yield) {
        seastar::thread::maybe_yield();
    }
    return boost::copy_range<inet_address_vector_replica_set>(tm.get_all_endpoints());
}
```
but this seems a bit hacky. And I don't see much point in yielding in `get_all_endpoints` too for the same reason (it's fast). I think `calculate_natural_endpoints` could easily be a non-async function which doesn't take `can_yield` parameter due to its speed. It's the functions which call this repeatedly in a loop that cause the problem, and these functions should be the ones yielding IMO. I.e.: you don't cause a problem (you don't perform a long computation) - you don't yield; you cause a problem, you must yield.

Reply:
It's the current design.
Maybe it isn't ideal, but it's better to be consistent with it or change it system-wide, not do a partial change just for this case.

Reply:
No point in yielding in the beginning, it's the responsibility of the caller in case the quota was already exceeded.

Agree that O(n) algorithms (for reasonable n) do not need yields.

Reply:
So I'm not sure what to do to make every reviewer happy.
>  Maybe it isn't ideal, but it's better to be consistent with it or change it system-wide, not do a partial change just for this case.

Well I'm not changing/removing existing yields, I'm just adding new yields on top.

Here are the options I see:
1. Yield at beginning/end of `everywhere_replication_strategy::calculate_natural_endpoints` <- but see Avi's concern (which I agree with)
2. Or modify `everywhere_replication_strategy::calculate_natural_endpoints` by expanding the loop that copies endpoints from `tm` and putting the yield inside the expanded loop <- but that's just obfuscating the logic with disputable gain, I much prefer the `boost::copy_range`
3. Or leave it as it is <- I'm in favor of this one; it seems that @bhalevy does not disagree with the reasoning that there is no point yielding inside fast algorithms, the only concern here is that it "mixes" two approaches. However, maybe it's good that we introduce this new approach into the code right now, as it will pressure us to continue going into that direction in later changes. If we keep doing things the old way then the old way will prevail.

Reply:
I guess there's also option
4. system-wide refactor so we go with the new approach (never yield inside fast algorithms - only functions which "really" introduce a problem must yield) <- yeah, but I don't see the value in doing this right now, the situation may change. I'd prefer fixing this bug with an easy patch right now. If we decide to do the refactor this PR would make the work a bit smaller so it's a step in the right direction nevertheless.

Reply:
What about

5. Agree that yielding should be done by slow functions, not fast functions, but do not require an overnight change.

Reply:
I think that would correspond to option 3.

Reply:
Practically, yes, but it also includes a statement that the status quo is wrong and we want to change it.

Reply:
And I guess you'd like this statement to be included in one of the comments somewhere (like a FIXME) or/and commit message?

Also, cc @bhalevy @asias - what do you think? You suggested to put yields in `calculate_natural_endpoints` and `token_metadata::get_all_endpoints` as well (probably in a followup), but if we go with option 3./5. discussed above, this would actually mean we go in a completely opposite direction - in particular, _remove_ `can_yield` parameter from `calculate_natural_endpoints` (because it's a linear algorithm... although I'm not 100% sure for NTS implementation, but it _should_ be linear, I know from a high level that it _can_ be) (removing `can_yield` from there could be done eventually, not necessarily in this PR). Are you fine with that?

Reply:
I'm certainly okay with making forward progress and not sticking mistakes we made in the past.
Meaning we can go ahead with the approach in this patch, documenting (with FIXME comment(s)), and opening github issue(s) to fix them and remove yielding from the lower level functions that have linear complexity.

Reply:
The n for calculate_natural_endpoints is number of nodes * number of tokens per node. This can be really big in large clusters, especially if user uses a much bigger  num_tokens (default is 256).  I am not so confident it is 100% stall free. 

To move forward, I am fine with this patch, since we use get_address_ranges in limited places.